File: prototype/project.rs
use std::collections::HashMap;
use may_actor::Actor;

use std::{rc::Rc, cell::RefCell};
use main::do_in_main;
use gtk::{self, prelude::*};
use gdk::enums::key;

use gallery::Gallery;
use editor::Editor;

pub enum Item {
  Editor,
  Gallery
}

use files_tree::FilesTree;
use web_browser::WebBrowser;

pub struct View {
  item: Item,
  files_tree: FilesTree,
  floating_layer: gtk::Container
}

impl View {
  fn new() {
    let  = ;
    let tree_view = ;
    let vbox = gtk::Box::new();
    vbox.pack();
    ProjectView { overlay, project_tree,  }
  }
}

pub struct Project {
  views: RefCell<HashMap<String, View>>,
  notebook: gtk::NoteBook,
  web_browsers: RefCell<HashMap<String, WebBrowser>>,
  external_projects: RefCell<HashMap<String, Project>>
}

impl Project {
  pub fn go_to_chapter(self, chapter_path: &str) {}
  pub fn prev_chapter(self) {}
  pub fn next_chapter(self) {}
}

File: prototype/project.js
const { Gio, Gdk, Gtk } = imports.gi;
const { Gallery } = imports.gallery;
const { Editor } = imports.editor;
const { FilesTree } = imports.files_tree;
const { WebBrowser } = imports.web_browser;

class View {
  // Gallery or Editor, FilesTree, Gtk.Container
  constructor(item, files_tree, floating_layer) {}
}

class Project {
  // [View], Gtk.NoteBook, { "url": WebBrowser }, { "url": Project }
  constructor(views, notebook, web_browsers, external_projects) {}

  go_to_chapter(chapter_path) {}

  prev_chapter() {}

  next_chapter() {}
}
